### ISSUE TITLE
Minimum cluster size with failover

### ISSUE THREAD

**Issue by ddorian** [NVIDIA/aistore#73]
Hi team,

Looking to create a small cluster, say, 3-4 nodes and with ability to grow.

Seems like you only replicate small objects with the same replication as parity-number. And a usual/popular EC scheme is at least 6+4, thus requiring 10 nodes. And EC can't be changed.

Is it possible to have only replication, until the cluster grows enough to handle, say, 10+4 EC ?

Or doing EC 10+4 from the start, with 4 nodes, and setting `ec.objsize_limit` to very big to cover all objects, and when the cluster is bigger, lowering it so big objects start being erasure-coded. 

Does that make sense ?

Regards

**Comment by VladimirMarkelov** [comment]
Hi,

thank you for the idea. As I understand, the proposal is:

1. Ability to set **any** `data:parity` value for a cluster
2. While the number of nodes is insufficient (AIS requires 11 nodes for `6+4`), the cluster only creates replicas of small objects (as many as possible but not greater than `parity`). As soon as the cluster grows up to 11 nodes, all new big objects get erasure-encoded, the old ones remains intact (if you need to erasure-code old objects, you can run `ais start ec-encode`)
3. It should be possible to change only one EC property `ec.objsize_limit` on the fly when EC is enabled for a bucket. The rest properties like data+parity cannot be changed on the fly(as it is done now)

Is it correct? If so, the extra question: what if the number of nodes is fewer than the number of parity slices? A cluster does nothing, or it creates as many replicas as the number of nodes in the cluster?

**Comment by ddorian** [comment]
@VladimirMarkelov 
Let me rephrase it. The only way to have a small cluster (3 or 4 nodes) is to use replication (for all objects). So:

1. Ability to start the cluster only with replication for all objects
2. Ability to convert from replication to erasure coding (after you grow the cluster, say, when you reach 10+ nodes, you want to convert to erasure coding)

My original question, was if we can achieve the functionality above^^ by using `ec.objsize_limit` with current feature set.

**Comment by ddorian** [comment]
> AIS requires 11 nodes for 6+4
I don't understand why the extra node is needed ?

**Comment by VladimirMarkelov** [comment]
> I don't understand why the extra node is needed ?

We keep a full replica on a single node. A bit more about it here: https://github.com/NVIDIA/aistore/blob/master/docs/storage_svcs.md#data-redundancy-summary-of-the-available-options-and-considerations

As of the original question, at this moment it is impossible to enable EC for a bucket if a cluster does not have a sufficient number of nodes to store all the slices and replicas. I checked the code and it seemed it should work fine without this restriction. After we test it thoroughly, we will publish a new version. Does it work for you?

**Comment by ddorian** [comment]
> We keep a full replica on a single node. A bit more about it here: https://github.com/NVIDIA/aistore/blob/master/docs/storage_svcs.md#data-redundancy-summary-of-the-available-options-and-considerations

Oh this is unacceptable. You should be able to read from ec-data, no need to keep duplicates.

**Comment by ddorian** [comment]
> In addition to Reed-Solomon encoded slices, we currently always store a full replica - the strategy that uses available capacity but pays back with read performance.

You can do file chunking or even create more data parts to get higher read speeds. This way your big-file will only be in 1 HDD so it'll actually end up being slower because of low hdd io/bandwidth.

**Comment by VladimirMarkelov** [comment]
In the master (commit `16cf9ae7`) we added a way to enable EC for a bucket if the number of targets is insufficient but enough to support N parity (for this a cluster needs N+1 targets). In this case, it requires `--force`: `ais set props ais://bck ec.enable true ec.data_slices 6 ec.parity_slices 4 --force`. 

The property `ec.objsize_limit` now can be modified on the fly even if the bucket is already erasure-encoded(with `--force` in the latter case): `ais set props ais://bck ec.objsize_limit 400000 --force`

> You can do file chunking or even create more data parts to get higher read speeds.

I agree it must be implemented. I was talking about the current requirements. We are going to make it optional soon.

**Comment by ddorian** [comment]
> I agree it must be implemented. I was talking about the current requirements. We are going to make it optional soon.

One way you can do chunking is implementing seaweedfs architecture, to enable efficient storage & small-files/chunks erasure coding with little overhead.

**Comment by VirrageS** [comment]
We will try to improve our EC implementation in the future.
